### ISSUE TITLE
Question: Debugging Main/Renderer Processes

### ISSUE THREAD

**Issue by gorphensu** [electron-react-boilerplate/electron-react-boilerplate#428]
how to debug main thread or node code?

i set 

``` json
"start-hot": "cross-env HOT=1 NODE_ENV=development electron --debug-brk=5858  -r babel-register -r babel-polyfill ./main.development",
```

but it does`t work ?
how can i do?

**Comment by amilajack** [comment]
Use the developer tools.

**Comment by wyjieying** [comment]
@amilajack  but the main.js file? how to debug?

**Comment by amilajack** [comment]
At the moment I'm not sure how to do this. We should make this a high priority.

**Comment by amilajack** [comment]
From what I've found, you need to setup an external debugger.

**Comment by aj-ptw** [comment]
I can't figure this out either

**Comment by amilajack** [comment]
Electron has a number of shortcomings when it comes to development experience. For me, two major issues are the developer tools not having fully-working integration (specifically the `debugger`) and the lack of tooling/discussion on debugging the main thread. But I think that eventually the community will provide idiomatic ways or built-in support for these issues. But at the moment, current solutions seem super tacky.

**Comment by Slapbox** [comment]
Setting up debugging for the main.js file is honestly more work than debugging blind. It's truly ridiculous. Wasted 6 hours and still couldn't get the debugging up. A solution would be great, but I'm sure it will not come easy.

**Comment by amilajack** [comment]
Lets merge this into https://github.com/chentsulin/electron-react-boilerplate/issues/379

**Comment by aj-ptw** [comment]
Yea i have spent numerous hours trying to get this to work too @Slapbox :/

**Comment by Slapbox** [comment]
@aj-ptw Your best bet it to just use console.log() statements as much as possible and refer to your terminal for the output.

Unless it's formally integrated into the boilerplate, I've just come to terms with the fact that I'm going to have to live without it.

**Comment by amilajack** [comment]
> Unless it's formally integrated into the boilerplate, I've just come to terms with the fact that I'm going to have to live without it.

A fix for this is being worked on at the moment.

**Comment by Slapbox** [comment]
@amilajack that will be awesome. I'm looking forward to it. Thanks as always for all your hard work.

**Comment by ritz078** [comment]
@amilajack any update about this feature ?

**Comment by aj-ptw** [comment]
so many console.log() hahah

**Comment by amilajack** [comment]
@ritz078 sorry for the delay on this. I've been looking into this for a while because its something I'm really excited for. I'll look into this over winter break

**Comment by aj-ptw** [comment]
@amilajack happy to look too, just have not been given much to work with but if you're able to get a PR with what your thinking I'm glad to jump in. I really really need this haha

**Comment by ritz078** [comment]
Guys I think this and hot reloading main process are the biggest bottleneck for developing project in this boilerplate. It turns out to be very irritating while development 😭

**Comment by amilajack** [comment]
@ritz078 would be great if you or anyone else could PR for this. At the moment, I dont have much time. But expect a PR for this from me during late december or early january if one hasnt been made already.

**Comment by panta82** [comment]
I got Webstorm's debugger working by removing webpack-specific features from the background process files and running `main.dev.js` directly with electron / node.

The features I lost:
- ES6 module syntax, have to use `require()`
- FlowType
- async/await (should be here pretty soon though)

All other ES6/ES7 features I'd use babel for seem already covered by node, so the tradeoff was worth it for me. The best part is, I can still use babel to build the optimized production version.

This is the setup:

![image](https://user-images.githubusercontent.com/4291324/27989760-0e2b53e0-6440-11e7-9084-4c25a8011799.png)

I start this configuration from Webstorm (with or without debugger), then run `npm run hot-updates-server` from a terminal. I can restart the main process freely as I make changes, or put it on nodemon, whatever.

IMO this is a good option for people who need to do a lot of development in the background process, or don't care about FlowType.

**Comment by Slapbox** [comment]
@panta82, very cool. Thanks for sharing

**Comment by DaanBroekhof** [comment]
Thanks @panta82! I got debug running on PhpStorm (is Webstorm) via your method.

For others attempting the same, I have made a branch on a fork that shows what I changed (not a lot):  https://github.com/chentsulin/electron-react-boilerplate/compare/master...DaanBroekhof:debug-main-process

In the current master, starting the renderer process is now `npm run start-renderer-dev` instead of hot-updates-server.

**Comment by aj-ptw** [comment]
I DONT CARE ABOUT FLOW TYPE AND DO A LOT OF DEVELOPMENT ON BACKGROUBD PROCESSES AND I USE WEBSTORM

Holllaaa thanks!!

**Comment by nessup** [comment]
I know this is a super old issue, but in case anyone stumbles here by way of Google: electron added main process debugging a while back. Practically, the way to debug the main process is to add `--inspect=5858` to `start-main-dev` in `package.json`. Since this was solved upstream, this issue can probably be closed.

**Comment by PitterL** [comment]
> Thanks @panta82! I got debug running on PhpStorm (is Webstorm) via your method.
> 
> For others attempting the same, I have made a branch on a fork that shows what I changed (not a lot): [master...DaanBroekhof:debug-main-process](https://github.com/electron-react-boilerplate/electron-react-boilerplate/compare/master...DaanBroekhof:debug-main-process)
> 
> In the current master, starting the renderer process is now `npm run start-renderer-dev` instead of hot-updates-server.

It's working by your method, do you know why need change the "import.. from..." to require() but original project don't need the change?

**Comment by iyobo** [comment]
I really don't think this should be closed as all that is being debugged is the renderer process, as opposed to Main and Renderer.

This works to debug the main 
`"start-main-dev": "cross-env HOT=1 NODE_ENV=development electron $NODE_DEBUG_OPTION -r @babel/register ./app/main.dev.js",`

The renderer process though is still not picking up on debugging even after
`"start-renderer-dev": "cross-env NODE_ENV=development webpack-dev-server $NODE_DEBUG_OPTION --config configs/webpack.config.renderer.dev.babel.js",`

**Comment by Do-do-new** [comment]
I had to browse through several posts (not only #428 and #528) until it worked for me with the current version, so here is a setup which worked in VS Code, at least with the current master (63cebabb467400c910acf51fa7a0b9285b7bff22)

In package.json add extra script:
    `"start-main-debug": "yarn start-main-dev --inspect=5858 --remote-debugging-port=9223"`

In launch.json:
```
{
  "version": "0.2.0",
  "configurations": [
      {
          "type": "node",
          "request": "launch",
          "name": "Electron: Main",
          "protocol": "inspector",
          "runtimeExecutable": "yarn",
          "runtimeArgs": [
              "start-main-debug"
          ],
          "preLaunchTask": "Start Webpack Dev"
      },
      {
        "name": "Electron: Renderer",
        "type": "chrome",
        "request": "attach",
        "port": 9223,
        "webRoot": "${workspaceFolder}",
        "timeout": 15000
      }
  ],
  "compounds": [
      {
          "name": "Electron: All",
          "configurations": [
              "Electron: Main",
              "Electron: Renderer"
          ]
      }
  ]
}
```

In targets.json:
```
{
  "version": "2.0.0",
  "tasks": [
    {
      "type": "npm",
      "label": "Start Webpack Dev",
      "script": "start-renderer-dev",
      "options": {
        "cwd": "${workspaceFolder}"
      },
      "isBackground": true,
      "problemMatcher": {
        "owner": "custom",
        "pattern": {
          "regexp": "____________"
        },
        "background": {
          "activeOnStart": true,
          "beginsPattern": "Compiling\\.\\.\\.$",
          "endsPattern": "(Compiled successfully|Failed to compile)\\.$"
        }
      }
    }
  ]
}

```

**Comment by pialin** [comment]
What @Do-do-new  said works for me, only if  I changed

> targets.json

to
 
>tasks.json

Awesome! Thank you!

**Comment by iyobo** [comment]
> Anybody have an idea on how to do this with Webstorm?

Webstorm hardly ever gets any love because it's paid.
I think it would be worth it at this point for JetBrains to slow vscode run settings be consumable by their IDE.

**Comment by ethanbonin** [comment]
> 
> 
> > Anybody have an idea on how to do this with Webstorm?
> 
> Webstorm hardly ever gets any love because it's paid.
> I think it would be worth it at this point for JetBrains to have vscode run settings be consumable by their IDE.

It's true. What a fantastic idea!

**Comment by ethanbonin** [comment]
Has anybody figured out how to connect it to the renderer? I used @Do-do-new file configuration, but it only working for the main process and not the renderer.

**Comment by AndreyDodonov-EH** [comment]
> Has anybody figured out how to connect it to the renderer? I used @Do-do-new file configuration, but it only working for the main process and not the renderer.

Are you using compound configuration (Electron: All) when starting? Okay, it is a stupid question, but anyway.
If yes, have you also tried debugging renderer-only?

**Comment by Do-do-new** [comment]
> Has anybody figured out how to connect it to the renderer? I used @Do-do-new file configuration, but it only working for the main process and not the renderer.

@ethanbonin  
Are you using compound configuration (Electron: All) when starting? Okay, it is a stupid question, but anyway.
If yes, have you also tried debugging renderer-only?

P.S. I will check the possibility to do it with Webstorm, even though it is in general not free, but it offers better refactoring support (at least, in my experience) and there is also EAP (early access program)

**Comment by ethanbonin** [comment]
@Do-do-new, yes. I have tried Electron-All. I actually have tried the single Electron-Renderer by itself. .

But just looking at the configuration. It’s just attaching itself to the chrome. So I don’t know how that would help without starting the electron app. 😳

**Comment by Do-do-new** [comment]
> @Do-do-new, yes. I have tried Electron-All. I actually have tried the single Electron-Renderer by itself. .
> 
> But just looking at the configuration. It’s just attaching itself to the chrome. So I don’t know how that would help without starting the electron app. 😳

@ethanbonin 
Unravel launch.json, the first part, Electron: Main. It has runtime argument, start-main-debug.
Which in turn calls start-main-dev. Which in turn starts electron app.
 (you can see that in package.json of the boilerplate).

**Comment by Do-do-new** [comment]
@ethanbonin  @iyobo 
As for the webstorm, also doable. The only problem is that I couldn't find a possibility to run "Before launch" task in backround, so you have to start renederer manually.
So.

1) run (and allow to keep running) `yarn start-renderer-dev`
2) Add two configurations, one `npm`, one `Attach to Node.js/Chrome`. 
npm should run start-main-debug (I prefer yarn, but configuration is still called npm)
Attaching should connect to 9223.
3) Start both of them clicking on debug icon. (for the attaching you will be asked to choose page to debug) 
    OR use compound configuration which will start both  
    That's it.

Here are screenshots and correspondent run configs (they are normally in .run as separate files, but they won't work without referencing in .idea/workspace.xml. so I suggest following the screenshots).
![image](https://user-images.githubusercontent.com/16781381/85746352-aee2b000-b706-11ea-8bf0-9945795b29d0.png)
![image](https://user-images.githubusercontent.com/16781381/85746297-a4c0b180-b706-11ea-9c88-664b3df85204.png)
![image](https://user-images.githubusercontent.com/16781381/85747604-b787b600-b707-11ea-85ba-7cc2defa4537.png)


```
<component name="ProjectRunConfigurationManager">
  <configuration default="false" name="Electron: Main" type="js.build_tools.npm">
    <package-json value="$PROJECT_DIR$/package.json" />
    <command value="run" />
    <scripts>
      <script value="start-main-debug" />
    </scripts>
    <node-interpreter value="project" />
    <envs />
    <method v="2" />
  </configuration>
</component>

<component name="ProjectRunConfigurationManager">
  <configuration default="false" name="Electron: Renderer" type="ChromiumRemoteDebugType" factoryName="Chromium Remote" port="9223" restartOnDisconnect="true">
    <method v="2" />
  </configuration>
</component>

<component name="ProjectRunConfigurationManager">
  <configuration default="false" name="Ectron: All" type="CompoundRunConfigurationType">
    <toRun name="Electron: Main" type="js.build_tools.npm" />
    <toRun name="Electron: Renderer" type="ChromiumRemoteDebugType" />
    <method v="2" />
  </configuration>
</component>
```

**Comment by ethanbonin** [comment]
@Do-do-new, that is pretty slick. Here's what I got. What I did is created the `Attach to Node.js/Chrome` configuration. I then added in the `Before Launch` the NPM script to run `start-main-debug`


![image](https://user-images.githubusercontent.com/20937454/85749159-39f38400-b6ce-11ea-9dbf-dafe9c85f026.png)

**Comment by Do-do-new** [comment]
@ethanbonin 
Is it working for you? 
Because, first, I don't see where you run renderer, second, I don't see how webstorm could attach to main here, and we discuss debugging both in parallel...

**Comment by ethanbonin** [comment]
@Do-do-new, You're totally right.

**Comment by ethanbonin** [comment]
@Do-do-new, I was able to get your main debug process working. The breakpoints work on the Main, but I can't seem to get it to work on the renderer. Any ideas?

**Comment by Do-do-new** [comment]
@ethanbonin 
Is your electron app displayed correctly? I mean, is renderer process running okay?
I ensure by running it explicitly (in case of webstorm) and with task in case of VS Code.
![image](https://user-images.githubusercontent.com/16781381/85752919-19e2b580-b70c-11ea-8e79-c4a54ab47dd9.png)

If yes, then you need to attach to it properly. When running compound configuration, the renderer part will ask you to select to what you want to attach. Select you app. For me it looks something like this:
![image](https://user-images.githubusercontent.com/16781381/85753402-8362c400-b70c-11ea-8f41-a3755955f5bb.png)

**Comment by ethanbonin** [comment]
@Do-do-new, When I run `yarn dev` or `yarn start-renderer-dev` without the debugger, everything works great. The main and the renderer both work and display the app. 

It gives me this weird bug when I try to run the **Electron: All** in debug and the renderer part never asks what to attach to. 


![image](https://user-images.githubusercontent.com/20937454/85756735-b38e7080-b6d4-11ea-9ba8-9136d506d674.png)

Here are my run configuration files: 
```XML
<!-- Electron Main-->
<component name="ProjectRunConfigurationManager">
  <configuration default="false" name="Electron: Main" type="js.build_tools.npm">
    <package-json value="$PROJECT_DIR$/package.json" />
    <command value="run" />
    <scripts>
      <script value="start-main-debug" />
    </scripts>
    <node-interpreter value="project" />
    <envs />
    <method v="2" />
  </configuration>
</component>

<!-- Electron Renderer-->
<component name="ProjectRunConfigurationManager">
  <configuration default="false" name="Electron: Renderer" type="ChromiumRemoteDebugType" factoryName="Chromium Remote" port="9223" restartOnDisconnect="true">
    <method v="2" />
  </configuration>
</component>

<!-- Electron All-->
<component name="ProjectRunConfigurationManager">
  <configuration default="false" name="Ectron: All" type="CompoundRunConfigurationType">
    <toRun name="Electron: Main" type="js.build_tools.npm" />
    <toRun name="Electron: Renderer" type="ChromiumRemoteDebugType" />
    <method v="2" />
  </configuration>
</component>
```

Just to ensure, here is my package.json scripts
```JSON
{
   "start-main-debug": "yarn start-main-dev --inspect=5858 --inspect-brk=9223 --remote-debugging-port=9223",
    "start-main-dev": "cross-env START_HOT=1 NODE_ENV=development electron -r ./internals/scripts/BabelRegister ./app/main.dev.ts",
    "start-renderer-dev": "cross-env NODE_ENV=development webpack-dev-server --config configs/webpack.config.renderer.dev.babel.js",
}
```

and I set inside **webpack.config.renderer.dev.babel.js** `devServer: {noInfo: false}` to false. Do you see anything I'm missing?

**Comment by Do-do-new** [comment]
@ethanbonin 
The only difference I see is that you use `--inspect-brk=9223` (don't know what it is though).

Funny, I now cloned currect version of the boilerplate and they update the setup with .vscode settings which are discussed here.
I will check if it works and create fork with webstorm settings.

**Comment by Do-do-new** [comment]
@ethanbonin
Just re-cloned boilerplate.
Current version (2e85c08438ef209ff9dcb636056628abb9f7aa99) also works without any extra magic (just what I described for Webstorm VS code works right away).
The only thing is that after yarn install I had to restart the PC, attaching wasn't working for some reason until I did so.

**Comment by prophetw** [comment]
> ### Debug both main and renderer
> I had to browse through several posts (not only #428 and #528) until it worked for me with the current version, so here is a setup which worked in VS Code, at least with the current master ([63cebab](https://github.com/electron-react-boilerplate/electron-react-boilerplate/commit/63cebabb467400c910acf51fa7a0b9285b7bff22))
> 
> In package.json add extra script: `"start-main-debug": "yarn start-main-dev --inspect=5858 --remote-debugging-port=9223"`
> 
> In webpack.config.renderer.dev.babel.js: Change `noInfo` in `devServer` from `true` to `false`
> 
> In launch.json:
> 
> ```
> {
>   "version": "0.2.0",
>   "configurations": [
>       {
>           "type": "node",
>           "request": "launch",
>           "name": "Electron: Main",
>           "protocol": "inspector",
>           "runtimeExecutable": "yarn",
>           "runtimeArgs": [
>               "start-main-debug"
>           ],
>           "preLaunchTask": "Start Webpack Dev"
>       },
>       {
>         "name": "Electron: Renderer",
>         "type": "chrome",
>         "request": "attach",
>         "port": 9223,
>         "webRoot": "${workspaceFolder}",
>         "timeout": 15000
>       }
>   ],
>   "compounds": [
>       {
>           "name": "Electron: All",
>           "configurations": [
>               "Electron: Main",
>               "Electron: Renderer"
>           ]
>       }
>   ]
> }
> ```
> 
> In targets.json:
> 
> ```
> {
>   "version": "2.0.0",
>   "tasks": [
>     {
>       "type": "npm",
>       "label": "Start Webpack Dev",
>       "script": "start-renderer-dev",
>       "options": {
>         "cwd": "${workspaceFolder}"
>       },
>       "isBackground": true,
>       "problemMatcher": {
>         "owner": "custom",
>         "pattern": {
>           "regexp": "____________"
>         },
>         "background": {
>           "activeOnStart": true,
>           "beginsPattern": "Compiling\\.\\.\\.$",
>           "endsPattern": "(Compiled successfully|Failed to compile)\\.$"
>         }
>       }
>     }
>   ]
> }
> ```

> Cooooool!   This works. 
> But in v4.3.0 we need change some files to make it work with vscode.

- First
> in v4.3.0 .erb/configs/webpack.config.base.ts
```ts
  stats: "errors-only"  # this should change to { all: undefined }
# webpack errors-only will only output error msg, no output like "Compiled successfully" will cause task never finished and vscode debugger will stun by this. 
```

- Second
In targets/tasks.json file endsPattern may diff in diff webpack version.
If endsPattern is not match, Start Webpack Dev will always running.
I test in webpack 5.60.0，my output is like 
- webpack 5.60.0 compiled successfully in 3037 ms
So I change it to below: 
```json
        "background": {
          "endsPattern": "(compiled successfully|failed to compile)"
        }
```
- Third
> in v4.3.0 .vscode/launch.json Electron: Main should change like your version put electron node args in package.json
```json
# in package.json
{
"start:main": "cross-env NODE_ENV=development electron --inspect=5858 --remote-debugging-port=9223 -r ts-node/register/transpile-only ./src/main/main.ts",
}
# in .vscode/launch.json
{
  "version": "0.2.0",
  "configurations": [
    {
      "name": "Electron: Main",
      "type": "node",
      "request": "launch",
      "protocol": "inspector",
      "runtimeExecutable": "npm",
      "runtimeArgs": ["start:main"],
      "preLaunchTask": "Start Webpack Dev"
    },
    {
      "name": "Electron: Renderer",
      "type": "chrome",
      "request": "attach",
      "port": 9223,
      "webRoot": "${workspaceFolder}",
      "timeout": 15000
    }
  ],
  "compounds": [
    {
      "name": "Electron: All",
      "configurations": ["Electron: Main", "Electron: Renderer"]
    }
  ]
}


```

**Comment by amilajack** [comment]
Reopening since the existing debugging solution is broken

**Comment by Evilu** [comment]
Any progress with debugging in webstorm?

**Comment by Evilu** [comment]
Easy fix for you Webstorm people:

**Node interpreter:** The usual one, in _node_modules/.bin/electron_

**Node Parameters**: _--require ts-node/register   ._ 

**Enviroment variables:** TS_NODE_TRANSPILE_ONLY=true;NODE_ENV=development

<img width="1038" alt="Screenshot 2023-02-09 at 15 53 35" src="https://user-images.githubusercontent.com/38602464/217942801-4dd88e34-ecfe-4568-b0c2-e508f98e8a34.png">


Click the _edit environment variables_ button for easy edit:
<img width="931" alt="image" src="https://user-images.githubusercontent.com/38602464/217943400-2cbf8b11-1680-4859-8e3e-1e95f40d7bbb.png">

<img width="589" alt="Screenshot 2023-02-09 at 15 54 38" src="https://user-images.githubusercontent.com/38602464/217943430-373e5b7a-d0f6-4624-8ba1-18bf26078743.png">


Now you'll be able to edit you main process easily.
